File: airbyte-commons-server/src/main/java/io/airbyte/commons/server/handlers/helpers/DestinationMatcher.java
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.commons.server.handlers.helpers;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import io.airbyte.api.model.generated.DestinationRead;
import io.airbyte.api.model.generated.DestinationSearch;
import org.apache.commons.lang3.StringUtils;

/**
 * Constructs a query for finding a query.
 */
public class DestinationMatcher implements Matchable<DestinationRead> {

  private final DestinationSearch search;

  public DestinationMatcher(final DestinationSearch search) {
    this.search = search;
  }

  @Override
  public DestinationRead match(final DestinationRead query) {
    if (search == null) {
      return query;
    }

    final DestinationRead fromSearch = new DestinationRead();
    fromSearch.name(StringUtils.isBlank(search.getName()) ? query.getName() : search.getName());
    fromSearch.destinationDefinitionId(search.getDestinationDefinitionId() == null ? query.getDestinationDefinitionId()
        : search.getDestinationDefinitionId());
    fromSearch
        .destinationId(search.getDestinationId() == null ? query.getDestinationId() : search.getDestinationId());
    fromSearch.destinationName(
        StringUtils.isBlank(search.getDestinationName()) ? query.getDestinationName() : search.getDestinationName());
    fromSearch.workspaceId(search.getWorkspaceId() == null ? query.getWorkspaceId() : search.getWorkspaceId());
    fromSearch.icon(query.getIcon());
    fromSearch.isVersionOverrideApplied(query.getIsVersionOverrideApplied());
    fromSearch.breakingChanges(query.getBreakingChanges());
    fromSearch.supportState(query.getSupportState());

    if (search.getConnectionConfiguration() == null) {
      fromSearch.connectionConfiguration(query.getConnectionConfiguration());
    } else if (query.getConnectionConfiguration() == null) {
      fromSearch.connectionConfiguration(search.getConnectionConfiguration());
    } else {
      final JsonNode connectionConfiguration = search.getConnectionConfiguration();
      query.getConnectionConfiguration().fieldNames()
          .forEachRemaining(field -> {
            if (!connectionConfiguration.has(field) && connectionConfiguration instanceof ObjectNode) {
              ((ObjectNode) connectionConfiguration).set(field, query.getConnectionConfiguration().get(field));
            }
          });
      fromSearch.connectionConfiguration(connectionConfiguration);
    }

    return fromSearch;
  }

}

File: airbyte-commons-server/src/main/kotlin/io/airbyte/commons/server/handlers/helpers/DestinationMatcher.kt
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.commons.server.handlers.helpers

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.node.ObjectNode
import io.airbyte.api.model.generated.DestinationRead
import io.airbyte.api.model.generated.DestinationSearch

/**
 * Utility for matching and merging destination search criteria with destination queries.
 * Allows partial updates by merging search parameters with existing destination data.
 */
class DestinationMatcher(
  private val search: DestinationSearch?,
) : Matchable<DestinationRead> {
  /**
   * Merges search criteria with the given destination query.
   * Search parameters take precedence over query values when provided.
   *
   * @param query the base destination to merge search criteria into
   * @return a new DestinationRead with merged values
   */
  override fun match(query: DestinationRead): DestinationRead {
    if (search == null) {
      return query
    }

    return DestinationRead().apply {
      name(search.name?.takeIf { it.isNotBlank() } ?: query.name)
      destinationDefinitionId(search.destinationDefinitionId ?: query.destinationDefinitionId)
      destinationId(search.destinationId ?: query.destinationId)
      destinationName(search.destinationName?.takeIf { it.isNotBlank() } ?: query.destinationName)
      workspaceId(search.workspaceId ?: query.workspaceId)
      icon(query.icon)
      isVersionOverrideApplied(query.isVersionOverrideApplied)
      isEntitled(query.isEntitled)
      breakingChanges(query.breakingChanges)
      supportState(query.supportState)
      resourceAllocation(query.resourceAllocation)

      val connectionConfiguration =
        when {
          search.connectionConfiguration == null -> query.connectionConfiguration
          query.connectionConfiguration == null -> search.connectionConfiguration
          else -> {
            val connCfg = search.connectionConfiguration

            query.connectionConfiguration.fieldNames().forEachRemaining { field ->
              if (!connCfg.has(field) && connCfg is ObjectNode) {
                connCfg.set<JsonNode>(field, query.connectionConfiguration[field])
              }
            }

            connCfg
          }
        }

      connectionConfiguration(connectionConfiguration)
    }
  }
}
